torch_xla/csrc/tensor_methods.cpp

```diff
@@ -2506,6 +2507,20 @@ std::tuple<XLATensorPtr, XLATensorPtr> topk(const XLATensorPtr& input,
       input->CreateFrom(torch::lazy::Value(node, 1), at::ScalarType::Long));
 }
 
+std::tuple<XLATensorPtr, XLATensorPtr> topk_symint(const XLATensorPtr& input,
+                                                   c10::SymInt k, int64_t dim,
+                                                   bool largest, bool sorted,
+                                                   bool stable) {
+  SymIntElements k_symint = SymIntElements(k);
+  torch::lazy::NodePtr node = torch::lazy::MakeNode<TopKSymInt>(
+      input->GetIrValue(), k_symint,
+      torch::lazy::GetCanonicalDimensionIndex(dim, input->shape().get().rank()),
+      largest, sorted, stable);
+  return std::make_tuple(
+      input->CreateFrom(torch::lazy::Value(node, 0)),
+      input->CreateFrom(torch::lazy::Value(node, 1), at::ScalarType::Long));
```

Comment:
Some pytorch tests require the static topk returned indices type to be int64, otherwise it throws an error in pytorch [here](https://github.com/pytorch/pytorch/blob/db6cfff827158c3f22a6707aed35d2013ec2d994/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L1960). I saw `nonzero` returns int32 indices. Why not just do a `int64 (pytorch) -> int32 (xla) -> int64 (pytorch)` conversion for handling indices?

Reply:
so what happened here is that underlying XLA type will still be `xla::s32`, however we mark the logical element type will be `at::ScalarType::Long`. This will create a problem since underlying s32 can not represent all numbers for long. We run into issues when we try to cast this kind of tensor(long logical type, real s32 type) to other dtypes, because `Cast` op doesn't know how to handle this kind of cases.

The right thing to do is probably do a manual s64->s32->s64 cast instead of playing the trick of `logical_element_type` here.

Reply:
Thanks for the clarification! Do we want to add something like `res = torch::lazy::MakeNode<Cast>(res, at::ScalarType::Long)` here or use
`xla::ConvertElementType(indices, xla::PrimitiveType::S64, /*device=*/nullptr))` in `CreateTopK` like before?

Reply:
Let's handle it inside `CreateTopK` so it is cleaner.